lib/sync/fullsync.go

```diff
@@ -0,0 +1,547 @@
+	StorageState       StorageState
+	TransactionState   TransactionState
+	BabeVerifier       BabeVerifier
+	FinalityGadget     FinalityGadget
+	BlockImportHandler BlockImportHandler
+	Telemetry          Telemetry
+	BlockState         BlockState
+	BadBlocks          []string
+	NumOfTasks         int
+	RequestMaker       network.RequestMaker
+}
+
+type Importer interface {
+	handle(*types.BlockData, BlockOrigin) (imported bool, err error)
+}
+
+type FullSyncStrategy struct {
+	requestQueue  *requestsQueue[*network.BlockRequestMessage]
+	unreadyBlocks *unreadyBlocks
+	peers         *peerViewSet
+	badBlocks     []string
+	reqMaker      network.RequestMaker
+	blockState    BlockState
+	numOfTasks    int
+	startedAt     time.Time
+	syncedBlocks  int
+	importer      Importer
+}
+
+func NewFullSyncStrategy(cfg *FullSyncConfig) *FullSyncStrategy {
+	if cfg.NumOfTasks == 0 {
+		cfg.NumOfTasks = defaultNumOfTasks
+	}
+
+	return &FullSyncStrategy{
+		badBlocks:  cfg.BadBlocks,
+		reqMaker:   cfg.RequestMaker,
+		blockState: cfg.BlockState,
+		numOfTasks: cfg.NumOfTasks,
+		importer:   newBlockImporter(cfg),
+		unreadyBlocks: &unreadyBlocks{
+			incompleteBlocks: make(map[common.Hash]*types.BlockData),
+			// TODO: cap disjoitChains to don't grows indefinitely
+			disjointChains: make([][]*types.BlockData, 0),
+		},
+		requestQueue: &requestsQueue[*network.BlockRequestMessage]{
+			queue: list.New(),
+		},
+		peers: &peerViewSet{
+			view:   make(map[peer.ID]peerView),
+			target: 0,
+		},
+	}
+}
+
+func (f *FullSyncStrategy) NextActions() ([]*syncTask, error) {
+	f.startedAt = time.Now()
+	f.syncedBlocks = 0
+
+	if f.requestQueue.Len() > 0 {
+		message, _ := f.requestQueue.PopFront()
+		return f.createTasks([]*network.BlockRequestMessage{message}), nil
+	}
+
+	currentTarget := f.peers.getTarget()
+	bestBlockHeader, err := f.blockState.BestBlockHeader()
+	if err != nil {
+		return nil, fmt.Errorf("while getting best block header")
+	}
+
+	// our best block is equal or ahead of current target.
+	// in the nodes pov we are not legging behind so there's nothing to do
+	if uint32(bestBlockHeader.Number) >= currentTarget {
+		return nil, nil
+	}
+
+	startRequestAt := bestBlockHeader.Number + 1
+	targetBlockNumber := startRequestAt + 128
+
+	if targetBlockNumber > uint(currentTarget) {
+		targetBlockNumber = uint(currentTarget)
+	}
+
+	ascendingBlockRequests := network.NewAscendingBlockRequests(startRequestAt, targetBlockNumber,
+		network.BootstrapRequestData)
+
+	return f.createTasks(ascendingBlockRequests), nil
+}
+
+func (f *FullSyncStrategy) createTasks(requests []*network.BlockRequestMessage) []*syncTask {
+	tasks := make([]*syncTask, len(requests))
+	for idx, req := range requests {
+		tasks[idx] = &syncTask{
+			request:      req,
+			response:     &network.BlockResponseMessage{},
+			requestMaker: f.reqMaker,
+		}
+	}
+	return tasks
+}
+
+func (f *FullSyncStrategy) IsFinished(results []*syncTaskResult) (bool, []Change, []peer.ID, error) {
```

Comment:
Would a better name for this function be something like `ProcessTaskResults`? I say this because the function really does a good amount more than just determining if the sync strategy is finished or not.

If I remember, the checking if finished logic isn't needed here since full sync never stops, but will be needed for other strategies. Maybe then you can rename the function, but have the boolean return value be called `IsFinished`

Reply:
renamed it to Process and added a comment